# School system, income and  venues data analysis of Boston

## 1. Introduction 
Boston is the capital and largest city of Massachusetts, United States. The city is the center of American higher education, health care and investment funds. It is one of the most educated cities in the United States. Its economic base is education, finance, medical care and technology. The City of Boston's Office of Neighborhood Services has designated 23 Neighborhoods as Allston, Back Bay, Bay Village, Beacon Hill, Brighton, Charlestown, Chinatown/Leather District, Dorchester, Downtown, East Boston, Fenway Kenmore, Hyde Park, Jamaica Plain, Mattapan, Mission Hill, North End, Roslindale, Roxbury, South Boston, South End, West End and West Roxbury. 
    
Due to the diversity of Boston, we would like to analyze the correlation between income and school district systems across the various districts of Boston, as well as the differences between these districts. We would also to analyze whether the characteristics of each district are related to their school district level and income level.
    
Our finding will be provided some information to people who want to rent a house in Boston to fit their lifestyle and education system they want.

## 2. Data
In this part, we will choose data from four parts.

I. <strong>The public school data of Boston.</strong> The data is obtained through https://www.kaggle.com/ndalziel/massachusetts-public-schools-data. This dataset compiles data from the following Massachusetts Department of Education reports:
Enrollment by Grade,
Enrollment by Selected Population,
Enrollment by Race/Gender,
Class Size by Gender and Selected Populations,
Teacher Salaries,
Per Pupil Expenditure,
Graduation Rates,
Graduates Attending Higher Ed,
Advanced Placement Participation,
Advanced Placement Performance,
SAT Performance,
MCAS Achievement Results,
Accountability Report.  In each case, the data is the latest available data as of August 2017.

II. <strong>The income data of Boston.</strong> The data is obtained through https://en.wikipedia.org/wiki/Boston. Data is from the 2008–2012 American Community Survey 5-Year Estimates. 

III. <strong>The Administrative District data of Boston. </strong> The data is obtained through https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/table/?refine.state=MA. This database was composed using ZIP code gazetteers from the US Census Bureau from 1999 and 2000. The latitude and longitude given for each ZIP code is typically (though not always) the geographic centroid of the ZIP code. 

IV. <strong>The most common venues data of Boston</strong>. Forsquare API will be used to get the most common venues of Boston. 

In [1]:
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
import requests
import csv

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 2.1 Public school data 

In [2]:
df1=pd.read_csv('MA_Public_Schools_2017.csv')

In [3]:
df1.head(5)

School Code                     School Name    School Type   Function  \
0        10505                   Abington High  Public School  Principal   
1        10003  Beaver Brook Elementary School  Public School  Principal   
2        10002        Center Elementary School  Public School  Principal   
3        10405            Frolio Middle School  Public School  Principal   
4        10015     Woodsdale Elementary School  Public School  Principal   

           Contact Name            Address 1 Address 2      Town State   Zip  \
0  Teresa Sullivan-Cruz   201 Gliniewicz Way       NaN  Abington    MA  2351   
1       Catherine Zinni  1 Ralph Hamlin Lane       NaN  Abington    MA  2351   
2        Lora Monachino   201 Gliniewicz Way       NaN  Abington    MA  2351   
3    Matthew MacCurtain   201 Gliniewicz Way       NaN  Abington    MA  2351   
4        Jonathan Hawes  128 Chestnut Street       NaN  Abington    MA  2351   

          Phone           Fax        Grade District Name  District Code  \
0  781-982-2160  781-982-0061  09,10,11,12      Abington          10000   
1  781-982-2185  781-982-2187  01,02,03,04      Abington          10000   
2  781-982-2195  781-982-0053         PK,K      Abington          10000   
3  781-982-2170  781-982-2173        07,08      Abington          10000   
4  781-982-2180  781-982-2184        05,06      Abington          10000   

   PK_Enrollment  K_Enrollment  1_Enrollment  2_Enrollment  3_Enrollment  \
0              0             0             0             0             0   
1              0             0           125           164           148   
2             69           134             0             0             0   
3              0             0             0             0             0   
4              0             0             0             0             0   

   4_Enrollment  5_Enrollment  6_Enrollment  7_Enrollment  8_Enrollment  \
0             0             0             0             0             0   
1           143             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0           159           169   
4             0           180           173             0             0   

   9_Enrollment  10_Enrollment  11_Enrollment  12_Enrollment  SP_Enrollment  \
0           124            109            123             92              4   
1             0              0              0              0              0   
2             0              0              0              0              0   
3             0              0              0              0              0   
4             0              0              0              0              0   

   TOTAL_Enrollment  First Language Not English  % First Language Not English  \
0               452                        24.0                           5.3   
1               580                        61.0                          10.5   
2               203                        26.0                          12.8   
3               328                        18.0                           5.5   
4               353                        21.0                           5.9   

   English Language Learner  % English Language Learner  \
0                      11.0                         2.4   
1                      42.0                         7.2   
2                       6.0                         3.0   
3                       6.0                         1.8   
4                       7.0                         2.0   

   Students With Disabilities  % Students With Disabilities  High Needs  \
0                        44.0                           9.7       130.0   
1                        77.0                          13.3       215.0   
2                        43.0                          21.2        69.0   
3                        43.0                          13.1       101.0   
4             

In [4]:
df1.shape

(1861, 302)

 Create a school dataframe, here we only use one paramenter <strong>"Progress and Performance Index (PPI) - All Students"</strong> to indicate the school quality. 

In [5]:
# define the dataframe columns
column_names = ['SchoolName','Town', 'Zip', 'PPI'] 

# instantiate the dataframe
school_df1 = pd.DataFrame(columns=column_names)

In [6]:
for index, row in df1.iterrows():
         school_df1 = school_df1.append({'SchoolName':df1.iloc[index]['School Name'],
                                      'Town': df1.iloc[index]['Town'],
                                      'Zip': '0'+str(df1.iloc[index]['Zip']),
                                     'PPI': df1.iloc[index]['Progress and Performance Index (PPI) - All Students']
                                        }, ignore_index=True)

In [7]:
school_df1.head()

SchoolName      Town    Zip   PPI
0                   Abington High  Abington  02351  76.0
1  Beaver Brook Elementary School  Abington  02351  69.0
2        Center Elementary School  Abington  02351   NaN
3            Frolio Middle School  Abington  02351  63.0
4     Woodsdale Elementary School  Abington  02351  65.0

In [8]:
school_df2=school_df1.groupby('Zip').mean()

In [9]:
school_df2.head()

PPI
Zip             
01001  58.400000
01002  62.666667
01005  57.000000
01007  65.500000
01010  83.000000

In [10]:
school_df2.reset_index(level=0, inplace=True)

In [11]:
school_df2.head(5)

Zip        PPI
0  01001  58.400000
1  01002  62.666667
2  01005  57.000000
3  01007  65.500000
4  01010  83.000000

In [12]:
school_df2.shape

(433, 2)

## 2.2 The income data of Boston

In [13]:
r = requests.get("https://en.wikipedia.org/wiki/Boston")
soup = BeautifulSoup(r.content, 'lxml')

 Scraping a Wikipedia table with Beautiful Soup

In [14]:
# Load article, turn into soup and get the tables.
tables = soup.find_all('table', class_='sortable')

In [15]:
# Search through the tables for the one with the headings we want.
for table in tables:
    ths = table.find_all('th')
    headings = [th.text.strip() for th in ths]
    if headings[:2] == ['Rank', 'ZIP code (ZCTA)']:
        break

In [16]:
csv_file=open('Boston_wiki_01.csv','w')
csv_writer=csv.writer(csv_file)
csv_writer.writerow(headings)

111

In [17]:
for item in table.tbody.find_all('tr'):
    text1=item.text
    data1=text1.split("\n")[1]
    data2=text1.split("\n")[3].split()[0]
    data3=text1.split("\n")[5]
    data4=text1.split("\n")[7]
    data5=text1.split("\n")[9]
    data6=text1.split("\n")[11]
    data7=text1.split("\n")[13]
    
    
    csv_writer.writerow([data1, data2, data3, data4, data5, data6, data7])

In [18]:
csv_file.close()

In [19]:
df2=pd.read_csv('Boston_wiki_01.csv')

In [20]:
df2.dropna(axis=0, inplace=True)
df2.shape

(31, 7)

In [21]:
df2.head(5)

Rank ZIP code (ZCTA)  Per capitaincome  Medianhouseholdincome  \
0  Rank             ZIP  Per capitaincome  Medianhouseholdincome   
1     1           02110          $152,007               $123,795   
2     2           02199          $151,060               $107,159   
3     3           02210           $93,078               $111,061   
4     4           02109           $88,921               $128,022   

   Medianfamilyincome  Population  Number ofhouseholds  
0  Medianfamilyincome  Population  Number ofhouseholds  
1            $196,518       1,486                  981  
2            $146,786       1,290                  823  
3            $223,411       1,905                1,088  
4            $162,045       4,277                2,190

In [22]:
df2.dtypes

Rank                     object
ZIP code (ZCTA)          object
Per capitaincome         object
Medianhouseholdincome    object
Medianfamilyincome       object
Population               object
Number ofhouseholds      object
dtype: object

We only use <strong>Medicanhouseholdincome</strong> as an indicator for income of Boston 

In [23]:
df2.drop([0], inplace=True)

In [24]:
df2.drop(columns=['Rank'], inplace=True)

In [25]:
df2.drop(columns=['Per capitaincome', 'Medianfamilyincome', 'Population', 'Number ofhouseholds'], inplace=True)

In [26]:
df2.rename(columns={'ZIP code (ZCTA)': "Zip"}, inplace=True)

In [27]:
df2 = df2.reset_index(drop=True)


In [28]:
df2.head()

Zip Medianhouseholdincome
0  02110              $123,795
1  02199              $107,159
2  02210              $111,061
3  02109              $128,022
4  02116               $87,630

In [29]:
df2.shape

(30, 2)

## 3.3 The Administrative District data of Boston

"https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/table/?refine.state=MA" downland data


In [30]:
df4=pd.read_csv('us-zip-code-latitude-and-longitude.csv')

In [31]:
df4.shape

(43191, 1)

In [32]:
df4.head(4)

Zip;City;State;Latitude;Longitude;Timezone;Daylight savings time flag;geopoint
49654;Leland;MI;45.023384;-85.75697;-5;1;45.023384                                         -85.756970                             
50450;Lake Mills;IA;43.418877;-93.52078;-6;1;43...                                         -93.520780                             
52594;Unionville;IA;40.833045;-92.6867;-6;1;40....                                         -92.686700                             
54307;Green Bay;WI;44.460064;-88.007382;-6;1;44...                                         -88.007382

In [33]:
df4.reset_index(level=0, inplace=True)
df4.head()

index  \
0  49654;Leland;MI;45.023384;-85.75697;-5;1;45.02...   
1  50450;Lake Mills;IA;43.418877;-93.52078;-6;1;4...   
2  52594;Unionville;IA;40.833045;-92.6867;-6;1;40...   
3  54307;Green Bay;WI;44.460064;-88.007382;-6;1;4...   
4   54812;Barron;WI;45.39701;-91.86337;-6;1;45.39701   

   Zip;City;State;Latitude;Longitude;Timezone;Daylight savings time flag;geopoint  
0                                         -85.756970                               
1                                         -93.520780                               
2                                         -92.686700                               
3                                         -88.007382                               
4                                         -91.863370

In [34]:
df4.shape

(43191, 2)

In [35]:
df4.drop(['Zip;City;State;Latitude;Longitude;Timezone;Daylight savings time flag;geopoint'], axis=1, inplace=True)

In [36]:
df4.head()

index
0  49654;Leland;MI;45.023384;-85.75697;-5;1;45.02...
1  50450;Lake Mills;IA;43.418877;-93.52078;-6;1;4...
2  52594;Unionville;IA;40.833045;-92.6867;-6;1;40...
3  54307;Green Bay;WI;44.460064;-88.007382;-6;1;4...
4   54812;Barron;WI;45.39701;-91.86337;-6;1;45.39701

In [37]:
#df4.iloc[1][0].split(';')

In [38]:
# define the dataframe columns
column_names = ['Zip','city', 'state', 'Latitude', 'Longitude'] 

# instantiate the dataframe
MA_df = pd.DataFrame(columns=column_names)

In [39]:
MA_df

Empty DataFrame
Columns: [Zip, city, state, Latitude, Longitude]
Index: []

In [40]:
item=0
for index, row in df4.iterrows():
    if (df4.iloc[index][0].split(';')[2] == 'MA'):
         MA_df = MA_df.append({'Zip':df4.iloc[index][0].split(';')[0],
                               'city': df4.iloc[index][0].split(';')[1],
                               'state': df4.iloc[index][0].split(';')[2],
                                'Latitude': float(df4.iloc[index][0].split(';')[3]),
                                'Longitude': float(df4.iloc[index][0].split(';')[4])}, ignore_index=True)

In [41]:
MA_df.head()

Zip           city state   Latitude  Longitude
0  01888         Woburn    MA  42.446396 -71.459405
1  02459  Newton Center    MA  42.321197 -71.193000
2  02454        Waltham    MA  42.356719 -71.250479
3  01068         Oakham    MA  42.349769 -72.044130
4  02702        Assonet    MA  41.792192 -71.064460

In [42]:
MA_df.shape

(766, 5)

## 3.4 Join tables 

### 3.4.1 Join school table and income table with the same zip code

In [43]:
school_df2.head()

Zip        PPI
0  01001  58.400000
1  01002  62.666667
2  01005  57.000000
3  01007  65.500000
4  01010  83.000000

In [44]:
school_df2.shape

(433, 2)

In [45]:
df2.head()

Zip Medianhouseholdincome
0  02110              $123,795
1  02199              $107,159
2  02210              $111,061
3  02109              $128,022
4  02116               $87,630

In [46]:
df2.shape

(30, 2)

In [47]:
df_join1=df2.join(school_df2.set_index('Zip'), on='Zip')

In [48]:
df_join1.head()

Zip Medianhouseholdincome   PPI
0  02110              $123,795   NaN
1  02199              $107,159   NaN
2  02210              $111,061   NaN
3  02109              $128,022   NaN
4  02116               $87,630  55.0

In [49]:
df_join1.shape

(30, 3)

### 3.4.3 Jion the MA_df coordinate table

In [50]:
df_join2=df_join1.join(MA_df.set_index('Zip'), on='Zip')
df_join2.head()

Zip Medianhouseholdincome   PPI    city state   Latitude  Longitude
0  02110              $123,795   NaN  Boston    MA  42.356532  -71.05365
1  02199              $107,159   NaN  Boston    MA  42.347130  -71.08234
2  02210              $111,061   NaN  Boston    MA  42.347974  -71.04463
3  02109              $128,022   NaN  Boston    MA  42.361477  -71.05417
4  02116               $87,630  55.0  Boston    MA  42.349622  -71.07372

In [51]:
df_join2.shape

(30, 7)

In [52]:
Boston_df=df_join2

In [53]:
Boston_df.drop(['city', 'state'], axis=1, inplace=True)

In [54]:
Boston_df.head(1)

Zip Medianhouseholdincome  PPI   Latitude  Longitude
0  02110              $123,795  NaN  42.356532  -71.05365

## Use geopy library to get the latitude and longitude values of Toronto.

In [55]:
address = 'Boston, MA'

geolocator = Nominatim(user_agent="boston_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Boston City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Boston City are 42.3602534, -71.0582912.


In [56]:
# create map of Boston using latitude and longitude values
map_Boston = folium.Map(location=[latitude, longitude], zoom_start=10)


    
map_Boston

Define Foursquare Credentials and Version

In [57]:
CLIENT_ID = '0UKYDUTATAWZOK1IPONWZ4VBIOHOXJOTUZXJTEIJSKRW3IAS' # your Foursquare ID
CLIENT_SECRET = 'OFUEVSLXLFHJONAEGDZGYE3Z34PKCRPMOVMOSA23CCWD1PX2' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 0UKYDUTATAWZOK1IPONWZ4VBIOHOXJOTUZXJTEIJSKRW3IAS
CLIENT_SECRET:OFUEVSLXLFHJONAEGDZGYE3Z34PKCRPMOVMOSA23CCWD1PX2


In [58]:
zip_latitude = Boston_df.loc[0, 'Latitude'] # neighborhood latitude value
zip_longitude = Boston_df.loc[0, 'Longitude'] # neighborhood longitude value

zip_name = Boston_df.loc[0, 'Zip'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(zip_name, 
                                                               zip_latitude, 
                                                               zip_longitude))

Latitude and longitude values of 02110 are 42.356532, -71.05365.


Now, let's get the top 100 venues that are in "02110" within a radius of 500 meters.

In [59]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    zip_latitude, 
    zip_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=0UKYDUTATAWZOK1IPONWZ4VBIOHOXJOTUZXJTEIJSKRW3IAS&client_secret=OFUEVSLXLFHJONAEGDZGYE3Z34PKCRPMOVMOSA23CCWD1PX2&v=20180605&ll=42.356532,-71.05365&radius=500&limit=100'

Send the GET request and examine the resutls

In [60]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d1284f86e46500039e7f26e'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Financial District',
  'headerFullLocation': 'Financial District, Boston',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 160,
  'suggestedBounds': {'ne': {'lat': 42.361032004500004,
    'lng': -71.04757177954377},
   'sw': {'lat': 42.3520319955, 'lng': -71.05972822045624}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5b05fb67f193c0002cbd9455',
       'name': 'Trillium Garden On The Greenway',
       'location': {'address': 'Atlantic Ave',
        'crossStreet': 'High St',
        'lat': 42.35653994232763,
   

In [61]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [62]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name   categories        lat        lng
0  Trillium Garden On The Greenway  Beer Garden  42.356540 -71.051100
1                    Kane's Donuts   Donut Shop  42.356209 -71.052895
2        The Rose Kennedy Greenway         Park  42.358227 -71.052055
3          Equinox Franklin Street          Gym  42.356074 -71.054484
4               Post Office Square         Park  42.356340 -71.055686

In [63]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))


100 venues were returned by Foursquare.


In [64]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip', 
                  'Zip Latitude', 
                  'Zip Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [65]:
Boston_venues = getNearbyVenues(names=Boston_df['Zip'],
                                   latitudes=Boston_df['Latitude'],
                                   longitudes=Boston_df['Longitude']
                                  )

02110
02199
02210
02109
02116
02108
02114
02111
02129
02467
02113
02132
02118
02130
02127
02135
02131
02136
02134
02128
02122
02124
02125
02163
02115
02126
02215
02119
02121
02120


In [66]:
print(Boston_venues.shape)
Boston_venues.head()

(1469, 7)


Zip  Zip Latitude  Zip Longitude                            Venue  \
0  02110     42.356532      -71.05365  Trillium Garden On The Greenway   
1  02110     42.356532      -71.05365                    Kane's Donuts   
2  02110     42.356532      -71.05365        The Rose Kennedy Greenway   
3  02110     42.356532      -71.05365          Equinox Franklin Street   
4  02110     42.356532      -71.05365               Post Office Square   

   Venue Latitude  Venue Longitude Venue Category  
0       42.356540       -71.051100    Beer Garden  
1       42.356209       -71.052895     Donut Shop  
2       42.358227       -71.052055           Park  
3       42.356074       -71.054484            Gym  
4       42.356340       -71.055686           Park

In [67]:
Boston_venues.groupby('Zip').count()

Zip Latitude  Zip Longitude  Venue  Venue Latitude  Venue Longitude  \
Zip                                                                          
02108           100            100    100             100              100   
02109           100            100    100             100              100   
02110           100            100    100             100              100   
02111           100            100    100             100              100   
02113           100            100    100             100              100   
02114            51             51     51              51               51   
02115            38             38     38              38               38   
02116           100            100    100             100              100   
02118            55             55     55              55               55   
02119             9              9      9               9                9   
02120            17             17     17              17               17   
02121            16             16     16              16               16   
02122            13             13     13              13               13   
02124            13             13     13              13               13   
02125            14             14     14              14               14   
02126            17             17     17              17               17   
02127            32             32     32              32               32   
02128            26             26     26              26               26   
02129            26             26     26              26               26   
02130            44             44     44              44               44   
02131            30             30     30              30               30   
02132            13             13     13              13               13   
02134            81             81     81              81               81   
02135            43             43     43              43               43   
02136            10             10     10              10               10   
02163            35             35     35              35               35   
02199           100            100    100             100              100   
02210            67             67     67              67               67   
02215            66             66     66              66               66   
02467            53             53     53              53               53   

       Venue Category  
Zip                    
02108             100  
02109             100  
02110             100  
02111             100  
02113             100  
02114              51  
02115              38  
02116             100  
02118              55  
02119               9  
02120              17  
02121              16  
02122              13  
02124              13  
02125              14  
02126              17  
02127              32  
02128              26  
02129              26  
02130              44  
02131              30  
02132              13  
02134              81  
02135              43  
02136              10  
02163              35  
02199             100  
02210              67  
02215              66  
02467              53

In [68]:
print('There are {} uniques categories.'.format(len(Boston_venues['Venue Category'].unique())))

There are 221 uniques categories.


Analyze Each Neighborhood

In [69]:
# one hot encoding
Boston_onehot = pd.get_dummies(Boston_venues[['Venue Category']], prefix="", prefix_sep="")

# add zip column back to dataframe
Boston_onehot['Zip'] = Boston_venues['Zip'] 

# move zip column to the first column
fixed_columns = [Boston_onehot.columns[-1]] + list(Boston_onehot.columns[:-1])
Boston_onehot = Boston_onehot[fixed_columns]

Boston_onehot.head()

Zip  ATM  Accessories Store  African Restaurant  Airport  \
0  02110    0                  0                   0        0   
1  02110    0                  0                   0        0   
2  02110    0                  0                   0        0   
3  02110    0                  0                   0        0   
4  02110    0                  0                   0        0   

   American Restaurant  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0                    0         0                 0            0           0   
1                    0         0                 0            0           0   
2                    0         0                 0            0           0   
3                    0         0                 0            0           0   
4                    0         0                 0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       0     0    0               0                 0   

   Bed & Breakfast  Beer Bar  Beer Garden  Belgian Restaurant  Big Box Store  \
0                0         0            1                   0              0   
1                0         0            0                   0              0   
2                0         0            0                   0              0   
3                0         0            0                   0              0   
4                0         0            0                   0              0   

   Bike Rental / Bike Share  Bistro  Board Shop  Boat or Ferry  Bookstore  \
0                         0       0           0              0          0   
1                         0       0           0              0          0   
2                         0       0           0              0          0   
3                         0       0           0              0          0   
4                         0       0           0              0          0   

   Boutique  Bowling Alley  Boxing Gym  Brazilian Restaurant  Breakfast Spot  \
0         0              0           0                     0               0   
1         0              0           0                     0               0   
2         0              0           0                     0               0   
3         0              0           0                     0               0   
4         0              0           0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0        0                0         0             0              0   
1        0                0         0             0              0   
2        0                0         0             0              0   
3        0                0         0             0              0   
4        0                0         0             0              0   

   Bus Station  Bus Stop  Café  Cambodian Restaurant  Caribbean Restaurant  \
0            0         0     0                     0                     0   
1            0         0     0                     0                     0   
2            0         0     0                     0                     0   
3            0         0     0       

In [70]:
Boston_onehot.shape


(1469, 222)

Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [71]:
Boston_grouped = Boston_onehot.groupby('Zip').mean().reset_index()
Boston_grouped

Zip       ATM  Accessories Store  African Restaurant   Airport  \
0   02108  0.000000           0.000000            0.000000  0.000000   
1   02109  0.000000           0.000000            0.000000  0.000000   
2   02110  0.000000           0.000000            0.000000  0.000000   
3   02111  0.000000           0.000000            0.000000  0.000000   
4   02113  0.000000           0.000000            0.000000  0.000000   
5   02114  0.000000           0.000000            0.000000  0.000000   
6   02115  0.000000           0.000000            0.000000  0.000000   
7   02116  0.000000           0.010000            0.000000  0.000000   
8   02118  0.018182           0.000000            0.018182  0.000000   
9   02119  0.000000           0.000000            0.111111  0.000000   
10  02120  0.000000           0.000000            0.058824  0.000000   
11  02121  0.000000           0.000000            0.000000  0.000000   
12  02122  0.000000           0.000000            0.000000  0.000000   
13  02124  0.000000           0.000000            0.000000  0.000000   
14  02125  0.000000           0.000000            0.000000  0.000000   
15  02126  0.000000           0.000000            0.000000  0.000000   
16  02127  0.000000           0.000000            0.000000  0.000000   
17  02128  0.000000           0.000000            0.000000  0.038462   
18  02129  0.000000           0.000000            0.000000  0.000000   
19  02130  0.000000           0.045455            0.000000  0.000000   
20  02131  0.000000           0.000000            0.000000  0.000000   
21  02132  0.000000           0.000000            0.000000  0.000000   
22  02134  0.000000           0.000000            0.000000  0.000000   
23  02135  0.000000           0.000000            0.000000  0.000000   
24  02136  0.000000           0.000000            0.000000  0.000000   
25  02163  0.000000           0.000000            0.000000  0.000000   
26  02199  0.000000           0.010000            0.000000  0.000000   
27  02210  0.000000           0.000000            0.000000  0.000000   
28  02215  0.000000           0.000000            0.000000  0.000000   
29  02467  0.000000           0.000000            0.000000  0.000000   

    American Restaurant  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0              0.050000      0.00          0.000000     0.000000    0.000000   
1              0.020000      0.00          0.000000     0.000000    0.000000   
2              0.060000      0.03          0.000000     0.000000    0.010000   
3              0.010000      0.00          0.000000     0.000000    0.000000   
4              0.010000      0.00          0.000000     0.000000    0.000000   
5              0.039216      0.00          0.000000     0.000000    0.000000   
6              0.026316      0.00          0.000000     0.026316    0.105263   
7              0.040000      0.00          0.000000     0.010000    0.000000   
8              0.018182      0.00          0.018182     0.000000    0.000000   
9              0.222222      0.00          0.000000     0.000000    0.000000   
10             0.000000      0.00          0.000000     0.058824    0.000000   
11             0.000000      0.00          0.000000     0.000000    0.000000   
12             0.000000      0.00          0.000000     0.000000    0.000000   
13             0.000000      0.00          0.000000     0.000000    0.000000   
14             0.000000      0.00          0.000000     0.000000    0.000000   
15             0.000000      0.00          0.000000     0.000000    0.000000   
16             0.031250      0.00          0.000000     0.000000    0.000000   
17             0.000000      0.00          0.000000     0.000000    0.000000   
18             0.038462      0.00          0.000000     0.000000    0.000000   
19             0.022727      0.00          0.000000     0.045455    0.000000   
20             0.100000      0.00          0.000000     0.000000    0.000000   
21           

In [72]:
Boston_grouped.shape

(30, 222)

 Print each neighborhood along with the top 5 most common venues

In [73]:
num_top_venues = 5

for hood in Boston_grouped['Zip']:
    print("----"+hood+"----")
    temp = Boston_grouped[Boston_grouped['Zip'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----02108----
                 venue  freq
0  American Restaurant  0.05
1       Sandwich Place  0.05
2          Pizza Place  0.04
3          Coffee Shop  0.04
4                Plaza  0.04


----02109----
                venue  freq
0  Italian Restaurant  0.22
1  Seafood Restaurant  0.11
2              Bakery  0.06
3       Historic Site  0.04
4                Park  0.04


----02110----
                 venue  freq
0   Seafood Restaurant  0.06
1          Coffee Shop  0.06
2  American Restaurant  0.06
3       Sandwich Place  0.05
4                Hotel  0.04


----02111----
                venue  freq
0  Chinese Restaurant  0.13
1    Asian Restaurant  0.08
2              Bakery  0.07
3    Sushi Restaurant  0.04
4         Coffee Shop  0.04


----02113----
                venue  freq
0  Italian Restaurant  0.39
1              Bakery  0.06
2         Pizza Place  0.05
3                Park  0.05
4  Seafood Restaurant  0.04


----02114----
                venue  freq
0  Italian Restaurant  0.0

write a function to sort the venues in descending order.¶


In [74]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


create the new dataframe and display the top 10 venues for each neighborhood.¶


In [75]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zip']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
zip_venues_sorted = pd.DataFrame(columns=columns)
zip_venues_sorted['Zip'] = Boston_grouped['Zip']

for ind in np.arange(Boston_grouped.shape[0]):
    zip_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Boston_grouped.iloc[ind, :], num_top_venues)

zip_venues_sorted.head()

Zip 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  02108   American Restaurant        Sandwich Place           Coffee Shop   
1  02109    Italian Restaurant    Seafood Restaurant                Bakery   
2  02110    Seafood Restaurant           Coffee Shop   American Restaurant   
3  02111    Chinese Restaurant      Asian Restaurant                Bakery   
4  02113    Italian Restaurant                Bakery                  Park   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Plaza           Pizza Place    Italian Restaurant   
1         Historic Site                  Park           Pizza Place   
2        Sandwich Place         Historic Site                 Hotel   
3               Theater           Coffee Shop      Sushi Restaurant   
4           Pizza Place           Coffee Shop    Seafood Restaurant   

  7th Most Common Venue  8th Most Common Venue    9th Most Common Venue  \
0            Restaurant             Steakhouse  New American Restaurant   
1                  Café                    Pub                    Hotel   
2                  Park            Salad Place               Steakhouse   
3                  Café  Performing Arts Venue                      Gym   
4                  Café         Sandwich Place                   Market   

  10th Most Common Venue  
0          Historic Site  
1         Sandwich Place  
2               Aquarium  
3               Dive Bar  
4            Yoga Studio

Run k-means to cluster the neighborhood into 5 clusters.

In [76]:
# set number of clusters
kclusters = 5

Boston_grouped_clustering = Boston_grouped.drop('Zip', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Boston_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 2, 0, 0, 2, 0, 0, 0, 0, 1], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [77]:
# add clustering labels
zip_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Boston_merged = Boston_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Boston_merged = Boston_merged.join(zip_venues_sorted.set_index('Zip'), on='Zip')

Boston_merged.head() # check the last columns!

Zip Medianhouseholdincome   PPI   Latitude  Longitude  Cluster Labels  \
0  02110              $123,795   NaN  42.356532  -71.05365               0   
1  02199              $107,159   NaN  42.347130  -71.08234               0   
2  02210              $111,061   NaN  42.347974  -71.04463               0   
3  02109              $128,022   NaN  42.361477  -71.05417               2   
4  02116               $87,630  55.0  42.349622  -71.07372               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0    Seafood Restaurant           Coffee Shop   American Restaurant   
1           Coffee Shop   American Restaurant        Clothing Store   
2            Restaurant                 Hotel           Coffee Shop   
3    Italian Restaurant    Seafood Restaurant                Bakery   
4                   Spa  Gym / Fitness Center        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Sandwich Place         Historic Site                 Hotel   
1                 Hotel        Ice Cream Shop                   Spa   
2   American Restaurant    Italian Restaurant            Donut Shop   
3         Historic Site                  Park           Pizza Place   
4   American Restaurant    Italian Restaurant                   Gym   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                  Park           Salad Place            Steakhouse   
1    Italian Restaurant    Seafood Restaurant                   Bar   
2      Asian Restaurant                   Bar            Steakhouse   
3                  Café                   Pub                 Hotel   
4        Cosmetics Shop           Coffee Shop         Jewelry Store   

  10th Most Common Venue  
0               Aquarium  
1           Dessert Shop  
2     Mexican Restaurant  
3         Sandwich Place  
4     Mexican Restaurant

In [78]:
Boston_merged['Cluster Labels'].value_counts()

0    20
4     6
2     2
3     1
1     1
Name: Cluster Labels, dtype: int64

In [79]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Boston_merged['Latitude'], Boston_merged['Longitude'], Boston_merged['Zip'], Boston_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 3. Methodology 
Methodology section which represents the main component of the report where you discuss and describe any exploratory data analysis that you did, any inferential statistical testing that you performed, and what machine learnings were used and why.


## 4. Results 
Results section where you discuss the results.


## 5. Discussion 
Discussion section where you discuss any observations you noted and any recommendations you can make based on the results.


## 6. Conclusion 
Conclusion section where you conclude the report.